shtools 4.7 (new formula)

MarkWieczorek opened the pull request:
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----
shtools is a Fortran 95 based library that is used for performing spherical harmonic transforms and related operations. It is the underlying motor of the python-based project pyshtools. More info can be found here: https://shtools.github.io/SHTOOLS/

SMillerDev commented on Formula/shtools.rb:
```diff
@@ -0,0 +1,44 @@
+
+  def caveats
+    <<~EOS
+      To use SHTOOLS with your gfortran code, compile with the options
```
Is this specific to homebrew?

MarkWieczorek replied:
No, these are the options that need to be included when you compile and link your code to `libSHTOOLS.a`.

SMillerDev replied:
In that case it should be in the upstream documentation, not in caveats.

MarkWieczorek replied:
Even though these instructions are found in the web documentation, a common question we always get is "what compiler flags do I need to use to link to libSHTOOLS.a and make it work?". If you don't think that this is appropriate, I won't complain too much about removing it.

SMillerDev replied:
It's really just for issues that are specific to the homebrew install (https://docs.brew.sh/Formula-Cookbook#caveats)

MarkWieczorek replied:
Would it be ok to note that the compiled modules and library are located in `/usr/local/include` and `/usr/local/lib`, respectively? This would be in contrast to `/opt/local/...` when using macports. If not, I will just remove the entire caveats block.

SMillerDev replied:
That's pretty standard for homebrew so I wouldn't bother.

SMillerDev commented on Formula/shtools.rb:
```diff
@@ -0,0 +1,44 @@
+        -I/usr/local/include -lSHTOOLS -lfftw3 -lm -framework accelerate -m64 -O3
+      Location of the example code and data:
+        /usr/local/share/shtools
+      To run the test suite:
```
No need to tell people how to use brew.

SMillerDev commented on Formula/shtools.rb:
```diff
@@ -0,0 +1,44 @@
+  depends_on "gcc"
+
+  def install
+    ENV.deparallelize
```
Why can't it build in parallel?

MarkWieczorek replied:
We are working on that. There is a complicated race situation occuring where it breaks 1/number-of-processors times. The easiest was to turn this off for now. It turns out that a new PR came in yesterday that might fix this, but I need to test it. This would likely be removed in the next release.

SMillerDev commented on Formula/shtools.rb:
```diff
@@ -0,0 +1,44 @@
+    system "make", "fortran-mp"
+
+    pkgshare.install "examples/fortran/", "examples/ExampleDataFiles/"
+    inreplace pkgshare/"fortran/Makefile", "../../lib", "/usr/local/lib"
```
This looks like it should be patched upstream.

MarkWieczorek replied:
The problem is that if decided to clone the git repo and manually compile the code yourself in your local folder, these files would not not located at `/usr/local`. These filenames/paths need to be changed when being "installed" elsewhere.

SMillerDev replied:
But that's what you provide `PREFIX`/`DESTDIR` for in the make command. So the software knows where it's going to be.

MarkWieczorek replied:
Ok. I'll refactor the Makefile to make it work that way when compiling locally.

Question: Should I close this PR, and re-open a new one when we release a minor update? Or should I just keep this open and modify it later?

SMillerDev replied:
Generally people modify it later

MarkWieczorek commented:
Hi @SMillerDev, I just added a revised version of this formula that makes use of the SHTOOLS 4.7.1 maintenance release that just came out. I think that it addresses all of your issues:

* The caveats only say how to run the test suite.
* Following modifications to our make files, we can now build this in parallel.
* Following upstream modifications to the makefiles, it is no longer necessary to use "inreplace"

If there is anything else you think should be changed, please let me know!

SMillerDev commented on Formula/shtools.rb:
```diff
@@ -0,0 +1,42 @@
+
+  def caveats
+    <<~EOS
+      To run the test suite: brew test shtools
```
This is documented in the brew manpage, no need to document this here.

SMillerDev commented on Formula/shtools.rb:
```diff
@@ -0,0 +1,42 @@
+                   "MODFLAG=-I/usr/local/include",
+                   "LIBPATH=/usr/local/lib",
+                   "LIBNAME=SHTOOLS",
+                   "FFTW=-L /usr/local/lib -lfftw3 -lm",
```
```suggestion
                   "FFTW=-L #{HOMEBREW_PREFIX}/lib -lfftw3 -lm",
```

SMillerDev commented on Formula/shtools.rb:
```diff
@@ -0,0 +1,42 @@
+                   "F95=gfortran",
+                   "F95FLAGS=-m64 -fPIC -O3 -std=gnu -ffast-math",
+                   "MODFLAG=-I/usr/local/include",
+                   "LIBPATH=/usr/local/lib",
```
```suggestion
                   "LIBPATH=#{HOMEBREW_PREFIX}/lib",
```

SMillerDev commented on Formula/shtools.rb:
```diff
@@ -0,0 +1,42 @@
+                   "run-fortran-tests-no-timing",
+                   "F95=gfortran",
+                   "F95FLAGS=-m64 -fPIC -O3 -std=gnu -ffast-math",
+                   "MODFLAG=-I/usr/local/include",
```
```suggestion
                   "MODFLAG=-I#{HOMEBREW_PREFIX}/include",
```

MarkWieczorek commented:
No problem with those suggestions!

MarkWieczorek commented:
It looks like the two Travis test failures above are unrelated to this new formula (it is in the setup homebrew step). Perhaps someone could restart these and see if we get a different result?

MarkWieczorek commented:
Please note that I needed to update the file hash as a result of this issue: https://github.com/python-versioneer/python-versioneer/issues/217

I always assumed that the hash of a file at GitHub could never change, but I was wrong!

SMillerDev commented:
Could you squash your commits into one matching the name of the PR?

MarkWieczorek commented:
No problem. I squashed the commits and the tests just passed.

SMillerDev reviewed (approved):
Thanks @MarkWieczorek  ! Without contributions like yours it'd be impossible to keep homebrew going with the high standards that users have come to expect from the project. You can feel good knowing that you've made the world a tiny bit better for homebrew users around the world! 👍 🎉
